src/game/world/GreedyMesher2d.ts

```diff
@@ -0,0 +1,127 @@
+import { Matrix2d } from '../util/Matrix2d'
+
+export interface Area<T> {
+  value: T
+  x: number
+  y: number
+  w: number
+  h: number
+  state: 'growingX' | 'growingY' | 'done'
+}
+
+export class GreedyMesher2d<T> {
+  public areas: Area<T>[]
+  public currentArea: Area<T>
+  public processed: Matrix2d<boolean>
+  public isDone: boolean
+
+  constructor(
+    private readonly width: number,
+    private readonly height: number,
+    private readonly dataGetter: (x: number, y: number) => T,
+    private readonly skip: (value: T) => boolean = (value) => !!value,
+    private readonly isSameValue: (a: T, b: T) => boolean = (a, b) => a === b
+  ) {
+    this.areas = []
+    this.currentArea = this.newArea(0, 0)
+    this.processed = new Matrix2d(width, height, false)
+    this.isDone = false
+  }
+
+  public call() {
+    while (this.step());
+
+    return this.areas
+  }
+
+  public step() {
+    if (this.currentArea.state === 'growingX') {
+      this.growX()
+    }
+    if (this.currentArea.state === 'growingY') {
+      this.growY()
+    }
+    if (this.currentArea.state === 'done') {
+      this.prepareNextArea()
+    }
```

Comment:
```suggestion
    else if (this.currentArea.state === 'growingY') {
      this.growY()
    }
    else {
      this.prepareNextArea()
    }
```

Reply:
I also decided not to use a else statement when the state is 'done' because it is more clear to the reader, in which case the next area gets prepared.

Reply:
I also decided not to use a else statement for the last condition, because it is more readable in this case. The reader knows exactly which state needs to be set in order for the next area to be prepared.

Reply:
Inspite of my earlier comment, I decided to update the control flow neither less. Instead of using a else statement as you proposed, I decided to throw an exception when the state is invalid. This will hopefully aid in debugging this feature once the something goes out of hand.